### PR TITLE
Test dashboard feature

### PR DESCRIPTION
**Pull request by asif-khan17**
**Fixed Issues**

#20915

**Files Changed**

TestsController.cs - GetAllTestAsync method changed to GetAllTest
TestsRepository.cs - GetAllTests method changed to GetAllTestsAsync
ITestRepository.cs - Extra space removed
TestRepositoryTest.cs - async await deployed
Test.cs - Regex Justified for Test name
test.module.ts - lint error solved 
test.routing.ts - lint error solved
test-dashboard.ts - lint error solved
tests.service.ts - lint error solved
test-dashboard.pipe.ts - lint error solved
tests-dashboard.component.ts - lint error solved and console.log removed
tests.component.ts - lint error solved
tests.model.ts - lint error solved

**Checks**

- [X] Naming Conventions
- [X] Server side code comments (proper English, grammar and no spelling mistakes)
- [X] Client side code comments (proper English, grammar and no spelling mistakes)
- [X] Unused variables, methods, blank spaces and name spaces. 
- [X] Optimal Code and code formatting
- [X] Commit History is proper, no merge is done. 
- [X] Commit/pull request messages should be proper to justify your feature
- [X] All test cases are executed provided by tester.

### CODE REVIEW

File: Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs [hunk 1]
```diff
@@ -40,11 +42,10 @@ public string CreateTest([FromBody] Test test)
         /// </summary>
         /// <returns>List of Tests</returns>
         [HttpGet]
-        public IActionResult GetAllTest()
+        public async Task<ActionResult> GetAllTest()
         {
-            var tests = _testRepository.GetAllTests();
+            var tests = await _testRepository.GetAllTests();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs]: await method name should suffix with an async word.

File: Trappist/src/Promact.Trappist.DomainModel/Models/Test/Test.cs [hunk 2]
```diff
@@ -7,6 +7,7 @@ public class Test : BaseModel
     {
         [Required] 
         [MaxLength(150, ErrorMessage = "Test Name length Should be less than 150")] 
+        [RegularExpression("^[a - zA - Z0 - 9_@ $#%&*^{}+;:<>()-]*$",ErrorMessage ="Enter A Valid Test Name")]
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.DomainModel/Models/Test/Test.cs]: please justify this regular expression.

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 3]
```diff
@@ -41,9 +42,9 @@ public bool UniqueTestName(Test test)
         /// Fetch all the tests from Test Model,Convert it into List
         /// </summary>
         /// <returns>List of Tests</returns>
-        public List<Test> GetAllTests()
-        {
-            var tests = _dbContext.Test.ToList();
+        public async Task<List<Test>> GetAllTests()
+        {     
+            List<Test> tests = await Task.Run(() => _dbContext.Test.ToList());
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: No need to create background thread. Please use proper methods.

File: Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs [hunk 4]
```diff
@@ -40,11 +41,10 @@ public string CreateTest([FromBody] Test test)
         /// </summary>
         /// <returns>List of Tests</returns>
         [HttpGet]
-        public IActionResult GetAllTest()
+        public async Task<ActionResult> GetAllTestAsync()
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs]: action name does not have async suffix

File: Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs [hunk 5]
```diff
@@ -20,7 +21,7 @@ public interface ITestsRepository
         /// Method defined to fetch Tests from Test Model
         /// </summary>
         /// <returns>List Of Tests</returns>
-        List<Test> GetAllTests();
+       Task<List<Test>> GetAllTestsAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Test/ITestsRepository.cs]: formatting of file is not proper.

File: Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs [hunk 6]
```diff
@@ -29,18 +29,18 @@ public TestsRepositoryTest(Bootstrap bootstrap)
...
         }
         /// <summary>
         /// Test Case For Emtpty Test Model
         /// </summary>
         [Fact]
         public void GetAllTestEmpty()
         {
-            var list = _testRepository.GetAllTests();
-            Assert.Equal(0, list.Count);
+            var list = _testRepository.GetAllTestsAsync();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Test/Tests/TestsRepositoryTest.cs]: async call should be awaitable

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/test-settings/test-settings.component.ts [hunk 7]
```diff
@@ -14,6 +14,7 @@ import { ActivatedRoute } from '@angular/router';
 export class TestSettingsComponent{
     testsettings: Test = new Test();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/test-settings/test-settings.component.ts]: initialization should go in constructor.

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/tests-dashboard/tests-dashboard.component.ts [hunk 8]
```diff
@@ -29,7 +32,7 @@ export class TestsDashboardComponent {
 
     //Get All The Tests From Server
     getAllTests() {
-        this.testService.getTests().subscribe((response) => { this.Tests = (response), console.log(this.Tests) });
+        this.testService.getTests().subscribe((response) => { this.Tests = (response); console.log(this.Tests); });
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/tests-dashboard/tests-dashboard.component.ts]: why console.log is written here?

File: Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs (hunk 4 above)
Comment (asif-khan17) [inline Trappist/src/Promact.Trappist.Core/Controllers/TestController.cs]: action name is GetAllTestAsync() it does have Async suffix

File: Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs [hunk 9]
```diff
@@ -30,8 +32,8 @@ public void CreateTest(Test test)
         public bool UniqueTestName(Test test)
         {
             var testObj = (from s in _dbContext.Test
-                          where s.TestName == test.TestName
-                          select s).FirstOrDefault();
+                           where s.TestName == test.TestName
+                           select s).FirstOrDefault();
```
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Repository/Test/TestsRepository.cs]: please write linq lambda queries

File: Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/test-settings/test-settings.component.ts (hunk 7 above)
Comment (Krunal-Promact) [inline Trappist/src/Promact.Trappist.Web/wwwroot/app/tests/test-settings/test-settings.component.ts]: not able to see any changes here regarding previous comment.

**Pull request merged**